File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int left = 0, right = 0, max = 0;

        for (char c : s.toCharArray()) {
            while (set.contains(c)) {
                set.remove(s.charAt(left++));
            }

            set.add(c);
            right++;
            max = Math.max(max, right - left);
        }

        return max;
    }
}

File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        val set = HashSet<Char>()
        var left = 0
        var right = 0
        var maxLen = 0

        for (c in s) {
            while (set.contains(c)) {
                set.remove(s[left++])
            }

            set.add(c)
            right++
            maxLen = maxOf(maxLen, right - left)
        }

        return maxLen
    }
}